components/webln/nwc.js

```diff
@@ -0,0 +1,224 @@
+    return new Promise(function (resolve, reject) {
+      (async function () {
+        // XXX set this to mock NWC relays
+        const MOCK_NWC_RELAY = false
+
+        // need big timeout since NWC is async (user needs to confirm payment in wallet)
+        const timeout = MOCK_NWC_RELAY ? 3000 : 60000
+        let timer
+        const resetTimer = () => {
+          clearTimeout(timer)
+          timer = setTimeout(() => {
+            sub?.close()
+            if (MOCK_NWC_RELAY) {
+              const heads = Math.random() < 0.5
+              if (heads) {
+                return resolve({ preimage: null })
+              }
+              return reject(new Error('mock error'))
+            }
+            return reject(new Error('timeout'))
+          }, timeout)
+        }
+        if (MOCK_NWC_RELAY) return resetTimer()
+
+        const payload = {
+          method: 'pay_invoice',
+          params: { invoice: bolt11 }
+        }
+        const content = await nip04.encrypt(secret, walletPubkey, JSON.stringify(payload))
+        const request = finalizeEvent({
+          kind: 23194,
+          created_at: Math.floor(Date.now() / 1000),
+          tags: [['p', walletPubkey]],
+          content
+        }, secret)
+        await relay.publish(request)
+        resetTimer()
+
+        const filter = {
+          kinds: [23195],
+          authors: [walletPubkey],
+          '#e': [request.id]
+        }
+        const sub = relay.subscribe([filter], {
+          async onevent (response) {
+            resetTimer()
+            try {
+              const content = JSON.parse(await nip04.decrypt(secret, walletPubkey, response.content))
+              if (content.error) return reject(new Error(content.error.message))
+              if (content.result) return resolve({ preimage: content.result.preimage })
+            } catch (err) {
+              return reject(err)
+            } finally {
+              clearTimeout(timer)
+              sub.close()
+            }
+          },
+          onclose (reason) {
+            clearTimeout(timer)
+            reject(new Error(reason))
+          }
+        })
+      })().catch(reject)
+    })
+  }, [relay, walletPubkey, secret])
+
+  const getInfo = useCallback(() => {
+    return new Promise(function (resolve, reject) {
+      (async function () {
+        const timeout = 5000
+        const timer = setTimeout(() => {
+          sub?.close()
+          reject(new Error('timeout'))
+        }, timeout)
+
+        const sub = relay.subscribe([
+          {
+            kinds: [13194],
+            authors: [walletPubkey]
+          }
+        ], {
+          onevent (event) {
+            clearTimeout(timer)
+            const supported = event.content.split()
+            resolve(supported)
+            sub.close()
+          },
+          // some relays like nostr.mutinywallet.com don't support NIP-47 info events
+          // so we simply check that we received EOSE
```

Comment:
Is this required anymore? Mutiny has switched to `relay.mutinywallet.com` for all new connections and that should behave properly now. AFAIK, we were the only hacked together relay that didn't respond correctly.

Reply:
> Mutiny has switched to relay.mutinywallet.com for all new connections and that should behave properly now. 

Yeah, works great now :)

However, I am not sure how other relays will behave in the future. Afaik, there are only two wallets that support NWC so far: Mutiny and Alby.

It's a trade-off between strictly enforcing spec compliance vs. hoping they just didn't implement the info event but do support `pay_invoice`.

But maybe I should indeed remove it for "fail fast" reasons (even if paying invoices might work).

Reply:
This was mostly a relay implementation problem, not a wallet service problem. NWC wallets _should_ be broadcasting the info event. We were previously but our relay was not responding to them. This was because we were using Blastr as our relay for these and that isn't really a proper relay. I doubt anyone else would be running custom relay software that doesn't do simple event type querying.

Reply:
Makes sense, removed in 8db4cd1